engine/table/src/main/java/io/deephaven/engine/table/impl/select/DynamicWhereFilter.java

```diff
@@ -241,79 +264,56 @@ public WritableRowSet filter(
-                .map(mp -> table.getColumnSource(mp.leftColumn())).toArray(ColumnSource[]::new);
-        final TupleSource<?> tupleSource = TupleSourceFactory.makeTupleSource(keyColumns);
-        return filterGrouping(selection, selectionIndexer, tupleSource);
-    }
-
-    private WritableRowSet filterLinear(RowSet selection, ColumnSource<?>[] keyColumns, TupleSource<?> tupleSource) {
-        if (keyColumns.length == 1) {
-            return filterLinearOne(selection, keyColumns[0]);
-        } else {
-            return filterLinearTuple(selection, tupleSource);
-        }
-    }
-
-    private WritableRowSet filterLinearOne(RowSet selection, ColumnSource<?> keyColumn) {
+    private WritableRowSet filterLinear(RowSet selection, TupleSource<?> tupleSource) {
         if (selection.isEmpty()) {
             return RowSetFactory.empty();
         }
 
         if (!kernelValid) {
-            setInclusionKernel = SetInclusionKernel.makeKernel(keyColumn.getChunkType(), liveValues, inclusion);
+            setInclusionKernel = SetInclusionKernel.makeKernel(tupleSource.getChunkType(), liveValues, inclusion);
             kernelValid = true;
         }
 
         final RowSetBuilderSequential indexBuilder = RowSetFactory.builderSequential();
 
-        try (final ColumnSource.GetContext getContext = keyColumn.makeGetContext(CHUNK_SIZE);
-                final RowSequence.Iterator rsIt = selection.getRowSequenceIterator()) {
-            final WritableLongChunk<OrderedRowKeys> keyIndices = WritableLongChunk.makeWritableChunk(CHUNK_SIZE);
-            final WritableBooleanChunk<Values> matches = WritableBooleanChunk.makeWritableChunk(CHUNK_SIZE);
+        final int maxChunkSize = (int) Math.min(CHUNK_SIZE, selection.size());
+        // @formatter:off
+        try (final ColumnSource.GetContext keyGetContext = tupleSource.makeGetContext(maxChunkSize);
+             final RowSequence.Iterator selectionIterator = selection.getRowSequenceIterator();
+             final WritableLongChunk<OrderedRowKeys> selectionRowKeyChunk =
+                     WritableLongChunk.makeWritableChunk(maxChunkSize);
+             final WritableBooleanChunk<Values> matches = WritableBooleanChunk.makeWritableChunk(maxChunkSize)) {
+            // @formatter:on
 
-            while (rsIt.hasMore()) {
-                final RowSequence chunkOk = rsIt.getNextRowSequenceWithLength(CHUNK_SIZE);
+            while (selectionIterator.hasMore()) {
+                final RowSequence selectionChunk = selectionIterator.getNextRowSequenceWithLength(maxChunkSize);
 
-                final Chunk<Values> chunk = Chunk.downcast(keyColumn.getChunk(getContext, chunkOk));
-                setInclusionKernel.matchValues(chunk, matches);
+                final Chunk<Values> keyChunk = Chunk.downcast(tupleSource.getChunk(keyGetContext, selectionChunk));
+                final int thisChunkSize = keyChunk.size();
+                setInclusionKernel.matchValues(keyChunk, matches);
```

Comment:
There isn't anything "special" about `SetInclusionKernel`; do we always know it will be an object chunk now?

There doesn't seem to be any value in copying into matches, to then just use it later as filter. I think we can simply check against `liveValues` as we iterate and build index.

Reply:
Because I combined the single column case and the tuple case, we do *not* know it will always be an object chunk. Set inclusion let's us do non-boxing inclusion tests. That is, it *is* special.